modules/object-collections.adoc

```diff
@@ -0,0 +1,29 @@
+// Module included in the following assemblies:
+//
+// * operating/create-use-collections.adoc
+:_content-type: CONCEPT
+[id="understanding-object-collections_{context}"]
+= Understanding object collections
+
+You can create an object collection that can be used across namespaces, clusters, and deployments. Some common use cases for collections include:
+
+* Creating a group of namespaces to identify an infrastructure area that is owned by a specific development team
+* Using a deployment label to group all of the deployments in a distributed application spanning multiple namespaces
+* Identifying an application used in used in multiple clusters, such as a development and production cluster, which requires the same policy regardless of location
+* Performing cluster mapping for an entire production or test environment
+* Using annotations to identify the owner of a vulnerability issue that impacts a range of deployments or images
+
+See the "Collections configuration examples" section for specific use case examples.
+
+You can define collections using both _explicit identifiers_, such as Cluster ID, Namespace ID, and Deployment ID and _implicit identifiers_, such as Kubernetes labels and annotations. Collections are dynamic, or resolved in runtime. Collections can be hierarchical and include other collections.
+
+[NOTE]
+====
+When upgrading to the PostgreSQL database option, {product-title-short} automatically generates collection objects based on the access scope definitions that were used in reports that you configured in {product-title-short} prior to the upgrade. {product-title-short} creates one or more collections for each configured report configuration. 
+====
+
+##Re: this comment on JIRA: "We should inform users about how we are generating these, what they will be called, and urge them to go through any generated ones to rename them to something more appropriate if desired." Question: Can someone give me more info about how we are generating them and the naming?##
```

Comment:
This would be Charmik or Oscar.

I think it might also be worthwhile to mention that there are cases where ACS cannot convert an access scope to an equivalent collection, and when this happens the user would need to manually edit the report to set a collection.

I can get you screenshots of what this would look like in the UI if that would be helpful since it is difficult to reproduce in staging.

Reply:
I'll leave it to Charmik to speak to this directly as he's the one who did the implementation on it. It's also probably worth noting though that the migration was done as a best effort and there are some possible Access Scope configurations that were possible that would not map directly to a created collection.
In the case that people had existing configurations they should be directed to go in and check their configurations to ensure they don't need to do any manual configurations on existing report configurations.

Reply:
When we create equivalent collection objects for access scopes that were used in report configurations, we will generate two or more collections for a single access scope, depending on the complexity of the access scope. The generated collections for a given access scope will be organized as follows:

1) Embedded collections : In order to mimic the exact selection logic of the access scope, we will need to generate one or more collections, combing whose matched deployments will result in the same selection of clusters and namespaces as was intended by the access scope. These collections will have names of the form "System-generated embedded collection i for the scope <access-scope-name>" where i would be a whole number starting from 0. Note that these embedded collections will not have any attached collections and just have cluster and namespace selection rules, but no deployment rules because the access scopes did not filter on deployments.

2) Root collection for the access scope : This is the collection that will be added to the report configuration/s. It will have a name of the form "System-generated root collection for the scope <acces-scope-name>". It will not define any rules, but will attach one or more embedded collections. The combination of these embedded collections will result in the same selection of clusters and namespaces as was intended by the access scope.

Also note that for access scopes that define cluster or namespace label selectors, we will only be able to migrate those scopes that have 'IN' operator between key and values. Access scopes with label selectors that were created using the ACS central UI will have used 'IN' operator by default. We do not support 'NOT_IN', 'EXISTS' and 'NOT_EXISTS' operators. For scopes that we cannot create collections for, we will add log messages during the migration. These logs will have the following form:

```
Failed to create collections for scope <scope-name>: Unsupported operator NOT_IN in scope's label selectors. Only operator 'IN' is supported.
The scope is attached to the following report configurations: [list of report configs]; Please manually create an equivalent collection and edit the listed report configurations to use this collection. Note that reports will not function correctly until a collection is attached.
```

Additionally, report configuration's info page in the UI will also indicate that such a report needs to attach a collection to it.

Let me know if you want to see a quick demo of the migration along with scopes that cannot be migrated and I will be able to schedule one.